below/tc/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+mod errors;
+mod types;
+
+#[cfg(test)]
+mod test;
+
+use std::collections::BTreeMap;
+
+use netlink_packet_core::{NetlinkHeader, NetlinkMessage, NetlinkPayload, NLM_F_DUMP, NLM_F_REQUEST};
+use netlink_packet_route::RouteNetlinkMessage;
+use netlink_packet_route::tc::TcMessage;
+use netlink_sys::constants::NETLINK_ROUTE;
+use netlink_sys::{Socket, SocketAddr};
+use nix::net::if_;
+
+pub use errors::*;
+pub use types::*;
+
+pub type TcStats = Vec<Tc>;
+pub type Result<T> = std::result::Result<T, TcError>;
+
+/// Get list of all `tc` qdiscs.
+pub fn tc_stats() -> Result<TcStats> {
+    let ifaces = get_interfaces()?;
+    read_tc_stats(ifaces, &get_netlink_qdiscs)
+}
+
+fn read_tc_stats(interfaces: BTreeMap<u32, String>, netlink_retriever: &dyn Fn() -> Result<Vec<TcMessage>>) -> Result<TcStats> {
+    let messages = netlink_retriever()?;
+    let tc_stats = messages
+        .into_iter()
+        .map(|msg| Tc::new(&interfaces, &msg))
+        .collect();
+
+    Ok(tc_stats)
+}
+
+/// Open a netlink socket to retrieve `tc` qdiscs.
+/// The underlying library sends a message of type `RTM_GETQDISC` to the kernel.
+/// The kernel responds with a message of type `RTM_NEWQDISC` for each qdisc.
+fn get_netlink_qdiscs() -> Result<Vec<TcMessage>> {
+    // open a socket
+    let socket = Socket::new(NETLINK_ROUTE).map_err(|e| TcError::Netlink(e.to_string()))?;
+    socket.connect(&SocketAddr::new(0, 0)).map_err(|e| TcError::Netlink(e.to_string()))?;
+
+    // create a netlink request
+    let mut nl_hdr = NetlinkHeader::default();
+    nl_hdr.flags = NLM_F_REQUEST | NLM_F_DUMP;
+    let msg = RouteNetlinkMessage::GetQueueDiscipline(TcMessage::default());
+    let mut packet = NetlinkMessage::new(nl_hdr, NetlinkPayload::from(msg));
+    packet.finalize();
+    let mut buf = vec![0; packet.header.length as usize];
+    packet.serialize(&mut buf[..]);
+
+    // send the request
+    socket.send(&buf[..], 0).map_err(|e| TcError::Netlink(e.to_string()))?;
+
+    // receive the response
+    let mut recv_buf = vec![0; 4096];
+    let mut offset = 0;
+    let mut response = Vec::new();
+    'out: while let Ok(size) = socket.recv(&mut &mut recv_buf[offset..], 0) {
+        loop {
+            let bytes = &recv_buf[offset..];
```

Comment:
two things:
1. why double borrow?
2. why keep track of offset? seems like you could always write-to and read-from the beginning of the buffer

Reply:
this is mostly inspired from this [example](https://github.com/rust-netlink/netlink-packet-route/blob/main/examples/dump_packet_links.rs) in `netlink-packet-route` library
from my understanding:
1. `recv()` accepts a mutable reference to `BufMut`, I believe that's why we need double borrow. it refuses to compile otherwise.
```rs
pub fn recv<B>(&self, buf: &mut B, flags: libc::c_int) -> Result<usize>
where
    B: bytes::BufMut,
{
    ...
}
```
i can change it to: `&mut recv_buf[offset..].borrow_mut()` instead of `&mut &mut recv_buf[offset..]`

2. right, if we pass an empty buffer (or even the same buffer) every time we can get rid of the offset.